dawgz/workflow.py

```diff
@@ -158,17 +164,10 @@ def ensure(self, condition: Callable) -> None:
 
     def require(self, condition: Callable) -> None:
         if self.array is None:
-            assert accepts(condition), 'precondition should not expect arguments'
-        else:
-            assert accepts(condition, 0), 'precondition should expect one argument'
-
-        self.preconditions.append(condition)
-
-    def reduce(self, conditions: List[Callable]) -> Callable:
-        if self.array is None:
-            return lambda: all(c() for c in conditions)
-        else:
-            return lambda i: all(c(i) for c in conditions)
+            assert accepts(condition, 0), 'precondition should not expect arguments'
+            self.preconditions.append(condition)
+        elif not accepts(condition, 0):
+            self.preconditions.append(lambda _: condition())
```

Comment:
This fails if `accepts(condition)` is also `False`.

 ```suggestion
        if self.array is None:
            assert accepts(condition), 'precondition should not expect arguments'
        elif accepts(condition):
            condition = lambda _: condition()
        else:
            assert accepts(condition, 0), 'precondition should expect at most one argument'
           
        self.preconditions.append(condition)
```

Reply:
Maybe it would be cleaner to change accept into `num_arguments` or some other name. The implementation would become:

```python
from inspect import getfullargspec

def n_arguments(f):
    return len(getfullargspec(f).args)  # This ignores *args and **kwargs
```

Then we can write

```python
assert n_arguments(condition) <= 1, 'precondition expects at most one argument'
```

Reply:
I think trying to parse the signature ourselves is dangerous because there are a lot of edge cases. For instance if the function has default values for the args, it is still callable without arguments. And default arguments are very useful because they allow to define jobs in a loop and "fixing" the iterator (you cannot use the iterator directly in the function because then it would be a reference and it would take the last value eventually.) Also, a function that gobbles every arguments (*args) is still valid.

With accepts, you are 100% sure that the signature match with the input (although I agree that the name "accepts" is not explicit.)

Reply:
Yes, exactly, that's why I restrained from implementing that solution.

Reply:
Speaking of that, should the condition to set `condition` to `lambda _: condition()` be `if not accepts(condition, 0)` or `if accepts(condition)` ? For the example you mentioned above (with a default value) this leads to a different result.